src/testhost.x86/TestHostTraceListener.cs

```diff
@@ -0,0 +1,93 @@
+            EqtTrace.Verbose("TestPlatformTraceListener.Setup: Added test platform trace listener.");
+
+#if NETCOREAPP2_1
+            try
+            {
+                // workaround for netcoreapp2.1 where the trace listener api is not called when 
+                // Debug.Assert fails. This method is internal, but the class is on purpose keeping the 
+                // callback settable so tests can set the callback
+                var field = typeof(Debug).GetField("s_ShowDialog", BindingFlags.Static | BindingFlags.NonPublic);
+                var value = field.GetValue(null);
+                field.SetValue(null, (Action<string, string, string, string>)ShowDialog);
+            }
+            catch (Exception ex)
+            {
+                EqtTrace.Error("TestPlatformTraceListener.Setup: Failed to replace inner callback to ShowDialog in Debug.Assert. Calls to Debug.Assert with crash the test host process. {0}", ex);
+            }
+#endif
+        }
+
+        public override void Write(string message)
+        {
+            throw GetException(message);
+        }
+
+        public override void WriteLine(string message)
+        {
+            throw GetException(message);
+        }
+
+        public static void ShowDialog(string stackTrace, string message, string detailMessage, string _)
+        {
+            var text = !string.IsNullOrEmpty(message)
+                ? !string.IsNullOrEmpty(detailMessage)
+                    ? (message + Environment.NewLine + detailMessage)
+                    : message
+                : null;
+            throw GetException(text);
+        }
+
+        private static DebugAssertException GetException(string message)
+        {
+            var debugTypes = new Type[] { typeof(Debug), typeof(Trace) };
+            var stack = new StackTrace(true);
+
+            var debugMethodFound = false;
+            var frameCount = 0;
+            MethodBase method = null;
+            foreach (var f in stack.GetFrames())
+            {
+                var m = f.GetMethod();
+                var declaringType = m?.DeclaringType;
+                if (!debugMethodFound && (declaringType == typeof(Debug) || declaringType == typeof(Trace)))
+                {
+                    method = m;
+                    debugMethodFound = true;
+                }
+
+                if (debugMethodFound)
+                {
+                    frameCount++;
+                }
+            }
+
+            var stackTrace = string.Join(Environment.NewLine, stack.ToString().Split(Environment.NewLine).TakeLast(frameCount));
+            var methodName = method != null ? $"{method.DeclaringType.Name}.{method.Name}" : "<method>";
+            var wholeMessage = $"Method {methodName} failed with '{message}', and was translated to { typeof(DebugAssertException).FullName } to avoid terminating the process hosting the test.";
+
+            return new DebugAssertException(wholeMessage, stackTrace);
```

Comment:
Does this mean that the user's test can catch `DebugAssertException`? Its probably ok, but I am not sure we have other examples of exceptions thrown from test platform that can be handled in the user's test (and test project will also need to reference MS.VS.TP.OM.dll to do that which is not normally something test projects need to reference).

Reply:
Yes, I can also make the exception internal, which was the original idea, but I am expecting that someone will want to catch this exception (and I should probably attach the oringinal message in the exception data), so we should put it in the correct namespace already, in case we need to publish it later. 

I've put it in ObjectModel which is one of the packages that we expect people to be able to build against, and it is a transitive dependency of Test.Sdk, which is added to the test project as part of our template. That way both "raw" tests can consume it, as well as adapters that wish to wrap this into their own style of `Assert`.

Reply:
Not sure - but this could be problematic since different versions of VS would ship with different versions of OM.dll. This means user may need a binding redirect to ensure that the type that they are compiled against matches the type that ends up being thrown at run time. Not sure if we already define binding redirect for all uses of OM.dll in dotnet.exe (or testhost.(x86)exe)...

Reply:
I don't know. Should I make the type internal for now so we can merge this?

Reply:
I made the exception type internal.